packages/jest-phabricator-coverage/example/JestUnitTestEngine.php

```diff
@@ -0,0 +1,173 @@
+     }
+     $test_result->setUserData($result['message']);
+     $results[] = $test_result;
+   }
+
+    return $results;
+  }
+
+  private function runCommands($commands) {
+    $futures = array();
+    foreach ($commands as $command) {
+      $bin = $command['bin'];
+      $options = implode(' ', $command['options']);
+      $paths = $command['paths'];
+      $futures[] = new ExecFuture("{$bin} {$options} %Ls", $paths);
+    }
+
+    $console = PhutilConsole::getConsole();
+
+    // Pass stderr through so we can give the user updates on test
+    // status as tests run.
+    $completed = array();
+    $iterator = new FutureIterator($futures);
+    foreach ($iterator->setUpdateInterval(0.2) as $_) {
+      foreach ($futures as $key => $future) {
+        if (isset($completed[$key])) {
+          continue;
+        }
+        if ($future->isReady()) {
+          $completed[$key] = true;
+        }
+        list(, $stderr) = $future->read();
+        $console->writeErr('%s', $stderr);
+        break;
+      }
+    }
+    // Finish printing output for remaining futures
+    foreach ($futures as $key => $future) {
+      if (!isset($completed[$key])) {
+        list(, $stderr) = $future->read();
+        $console->writeErr('%s', $stderr);
+      }
+    }
+    $results = array();
+    foreach ($futures as $future) {
+      $results[] = $this->getFutureResults($future);
+    }
+
+    return call_user_func_array('array_merge', $results);
+  }
+
+  private function runJSTests() {
+    $console = PhutilConsole::getConsole();
+    $root = $this->getRoot();
+
+    $result_arrays = [];
+    $paths = $this->getPaths();
+    $jest_paths = array();
+    foreach ($paths as $path) {
+      $ext = idx(pathinfo($path), 'extension');
+      if ($ext === 'js' || $ext === 'json') {
+        // Filter deleted modules because Jest can't do anything with them.
+        if (file_exists("$root/$path")) {
+          $jest_paths[] = "$root/$path";
+        }
+      }
+    }
+
+    $commands = [];
+    if (count($jest_paths) > 0) {
+      $console->writeOut("Running JavaScript tests.\n");
+      $commands[] = array(
+        'bin' => self::JEST_PATH,
+        'options' => $this->getJestOptions(),
+        'paths' => $jest_paths,
+      );
+    }
+
+    try {
+      $result_arrays[] = $this->runCommands($commands);
+    } catch (Exception $e) {
+      // Ignore the exception in case of failing tests
+      // As Jest should have already printed the results.
+      $result = new ArcanistUnitTestResult();
+      $result->setName('JavaScript tests');
+      $result->setResult(ArcanistUnitTestResult::RESULT_FAIL);
+      $result->setDuration(0);
+      $result_arrays[] = array($result);
+    }
+
+    $console->writeOut("Finished tests.\n");
+    return call_user_func_array('array_merge', $result_arrays);
+  }
+
+  private function getJestOptions() {
+    $output_JSON = $this->getOutputJSON();
+    $options = array(
+      '--colors',
+      '--findRelatedTests',
```

Comment:
if you're running the related files you probably want to `collectCoverageFrom` the same set of files, otherwise you'll generate coverage for everything that's required